### ISSUE TITLE
Doc request: how to do explicit invocations with multiple factors

### ISSUE THREAD

**Issue by sirosen** [tox-dev/tox#2766]
## What's the problem this feature will solve?

Under tox v3, I was able to pass a string with multiple factors to `tox -e`.
For example, in a project with a `tomli` factor to control the installation of `tomli` for TOML support on py<3.11, the following invocation was valid in v3:
```
tox -e 'py{37,310}{,-tomli}'
```

The above expression would expand to an environment list in the same way that those factors would expand in the `envlist`.

This is particularly useful for declaring CI builds and trying to run a large test matrix efficiently.
A simple project with one factor may have an envlist which reads with multiple python versions and the factor enabled/disabled:
```
envlist = py{37,38,39,310,311}{,-tomli}
```
and in CI, `tox` may be invoked with `tox -e 'py{,-tomli}`.

I have tried to reimplement this kind of logic with tox v4 using labels and `tox r -m`, but for mildly complex cases it becomes very verbose.

The following usage fails with a `ValueError`:
```
tox r -e 'py37{,-tomli}'
```
As does `tox -m ci` with the config
```
labels =
    ci = py37{,-tomli}
```

## Describe the solution you'd like

On the assumption that this is possible and I simply haven't figured it out yet
- if someone could share the solution here, that would help
- let's get it into the documentation as part of the tox4 vs tox3 differences (I'm happy to open a PR once I know how to do this)

## Alternative Solutions

On the assumption that this is no longer possible and it isn't considered a desirable feature
- let's document that it's not supported anymore on the tox4 vs tox3 differences page
- some reasonable usage should be defined which covers this kind of usage, and offered in docs as an alternative

I think at a minimum, the docs should say that you used to be able to write `tox -e 'py{,-foo}'` and now you need to have an explicit `labels` config.

## Additional context

For a concrete example of a project where this broke, and the change which fixes it, refer to
https://github.com/python-jsonschema/check-jsonschema/pull/204

**Comment by gaborbernat** [comment]
Did you try `tox r -f py37 -f tomli'?

**Comment by gaborbernat** [comment]
As we do say in https://tox.wiki/en/latest/cli_interface.html#tox-run--f

**Comment by sirosen** [comment]
I didn't get the kind of control with `-f` that I was looking for, but I may be misunderstanding it or how it interacts with the envlist.
```
$ tox r -f py37 -f tomli
...
  py37-tomli: OK (10.20=setup[1.60]+cmd[8.59] seconds)
  py37-tomli-format: OK (9.87=setup[0.99]+cmd[8.89] seconds)
  py310-tomli: OK (10.97=setup[1.50]+cmd[9.47] seconds)
  py310-tomli-format: OK (10.70=setup[0.95]+cmd[9.75] seconds)
  py-tomli: OK (13.93=setup[1.36]+cmd[12.58] seconds)
  py-tomli-format: OK (13.93=setup[1.17]+cmd[12.77] seconds)
```

I wanted to get the environments of `py37` and `py37-tomli`. The other environments surprised me.

The envlist from my `tox.ini` for that run:
```
envlist =
    cov_clean
    py{311,310,39,38,37}
    py{37,310}-{tomli}{,-format}
    py{37,311}-{json5,pyjson5}
    cov
```

So I'm guessing that `-f tomli`  selects all environments from the envlist with `tomli` as a factor?
But that's not really what I want in this case, since the usage pattern is that I'm trying to select some precise slice of my test matrix.

It also seems a little odd that `py-tomli` would be selected, when the invocation specified `py37`.

I need to look at this config a bit more. Maybe there's a way of phrasing it such that `-f` selection would work for me.

---

It's okay if the kind of usage that I was doing under v3 isn't available anymore -- I can and will adjust. 🙂 
I'm just trying to understand what the new desired way is to cut a slice out of a larger test matrix.

**Comment by gaborbernat** [comment]
I'd accept a PR that allows an and operation instead of or for factor selection. Please submit the PR if you want this feature. What you want currently is not possible.

**Comment by sirosen** [comment]
Sorry about the long delay without a reply, but I now have this working and will put in a PR shortly.

"AND" factor selection is a great solution for this, and the implementation isn't very complex at all (especially with the pointer to where I had to add it! thanks!).

Merely having doc that `-f foo-bar` requires a match on `foo AND bar` will help to clarify that `-f foo -f bar` means `foo OR bar`. That was what I originally found surprising about factor selection.

**Comment by sirosen** [comment]
I was trying to add the documentation for this after my last comment and I realized that the logic today isn't quite what I thought it was.

`-f foo -f bar` does not mean "foo OR bar". It just means the same thing as `-f bar` -- I was imagining parsing with behavior like argparse's `action="append"`, but it's really using `nargs="+"`.

The logic today is already "AND". `-f foo bar` means `foo AND bar`.

What I had added in my branch for handling `-f foo-bar` is somewhat different, but given that it's hard to explain the difference I don't think the logic is good to add.

I am now looking at other possible ways to make factors behave. I _think_ the best path forward is to allow multiple uses of `-f` with "OR" semantics: `-f foo -f bar` means `foo OR bar` and `-f foo -f bar baz` means `foo OR (bar AND baz)`.

I'm not sure how adding `action="append"` plus `nargs="+"` behaves, but I'm about to start tinkering with this.
